
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Lab &#8212; Quantitative Reasoning in Biology</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_bblearn/Module02/Module02_lab';</script>
    <link rel="canonical" href="https://damlabresources.github.io/quantitative-reasoning-in-biology/_bblearn/Module02/Module02_lab.html" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Module 3: DataFrames" href="../../content/Module03/Module03_book.html" />
    <link rel="prev" title="Walkthrough" href="Module02_walkthrough_SOLUTION.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../content/book_index.html">
  
  
  
  
  
  
    <p class="title logo__title">Quantitative Reasoning in Biology</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../content/book_index.html">
                    Quantitative Reasoning in Biology
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Part 0: Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../content/misc/how_to.html">How to use this book</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Part 1 : Hello World</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../content/Module01/Module01_book.html">Module 1: Hello World</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../content/Module01/notebook_actions.html">Notebook basics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../content/Module01/Module01_walkthrough.html">Walkthrough</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../../content/Module02/Module02_book.html">Module 2: Simple calculations</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../content/Module02/dilution_calculations.html">Molarity Calculations</a></li>
<li class="toctree-l2"><a class="reference internal" href="Module02_walkthrough_SOLUTION.html">Walkthrough</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Lab</a></li>


</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../content/Module03/Module03_book.html">Module 3: DataFrames</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../content/Module03/spreadsheet_organization.html">Spreadsheet Organization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Module03/Module03_walkthrough_SOLUTION.html">Walkthrough</a></li>

<li class="toctree-l2"><a class="reference internal" href="../Module03/Module03_lab.html">Lab</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Part 2 : Descriptive Statistics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../content/Module04/Module04_book.html">Module 4: Analysis by groups</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../Module04/Module04_walkthrough_SOLUTION.html">Walkthrough</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../content/Module04/reshaping_data.html">Reshaping Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Module04/Module04_lab.html">Lab</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../content/Module05/Module05_book.html">Module 5: Plotting with Pandas</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../Module05/Module05_walkthrough_SOLUTION.html">Walkthrough</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Module05/Module05_lab.html">Lab</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../content/Module06/Module06_book.html">Module 6: Visualizing with Confidence</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../content/Module06/grammar_of_graphics.html">Grammar of Graphics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Module06/Module06_walkthrough_SOLUTION.html">Walkthrough</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Module06/Module06_lab.html">Lab</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../content/Module07/Module07_book.html">Module 7: Samples and Replicates</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../Module07/Module07_walkthrough_SOLUTION.html">Walkthrough</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Module07/Module07_lab.html">Lab</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Part 3 : Inferential Statistics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../content/Module08/Module08_book.html">Module 8: Hypothesis Testing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../Module08/Module08_walkthrough_SOLUTION.html">Walkthrough</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Module08/Module08_lab.html">Lab</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../content/Module09/Module09_book.html">Module 9: Linear Regression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../Module09/Module09_walkthrough_SOLUTION.html">Walkthrough</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Module09/Module09_lab.html">Lab</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../content/Module10/Module10_book.html">Module 10: Power Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../Module10/Module10_walkthrough_SOLUTION.html">Walkthrough</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Module10/Module10_lab.html">Lab</a></li>

</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Extra</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../content/misc/about_this_book.html">About this book</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../content/misc/course_info.html">Course Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../content/misc/author_bio.html">About the Author</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../content/misc/run_your_own_jupyter.html">Running Your Own Jupyter Environment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../content/faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../content/changelog.html">Change Log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../content/misc/works_cited.html">Works Cited</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/_bblearn/Module02/Module02_lab.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Lab</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Lab</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-objectives">Learning Objectives</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#protocol-evaluation">Protocol Evaluation</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#q1-extract-the-relevant-information-from-the-text-above">Q1: Extract the relevant information from the text above</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#q2-calculate-the-molecular-weight-of-each-template">Q2: Calculate the molecular weight of each template</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#q3-what-is-the-molarity-of-each-paragon-sample">Q3: What is the <em>molarity</em> of each <em>Paragon</em> sample?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#q4-what-is-the-yield-of-each-pacbio-sample">Q4: What is the yield of each <em>PacBio</em> sample?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#q5-which-samples-are-usable">Q5: Which samples are <em>usable</em>?</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#conclusion">Conclusion</a></li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="lab">
<h1>Lab<a class="headerlink" href="#lab" title="Link to this heading">#</a></h1>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Link to this heading">#</a></h2>
<p>In this lab, we will delve into the exciting world of viral sample preparation for nanopore sequencing for COVID-19 sequencing.
Specifically, we will be comparing the yield of two different RT-PCR methods: the Paragon system and the PacBio system.</p>
<p>Before we can sequence our samples, we need to make sure that we have a sufficient amount of DNA to work with.
This means calculating the appropriate dilutions of our samples to ensure that we have enough material to generate accurate and reliable results.</p>
<p>The Paragon system uses over 200 pairs of overlapping PCR primers that generate DNA fragments that are, on average, 285 basepairs long.
On the other hand, the PacBio system uses 35 overlapping fragments that have an average length of 2200 bp.
It will be interesting to see which method produces a higher yield of DNA and is more efficient in preparing our samples for sequencing.</p>
<p>Get ready to dive into some math and use Python to compare these two RT-PCR methods!</p>
</section>
<section id="learning-objectives">
<h2>Learning Objectives<a class="headerlink" href="#learning-objectives" title="Link to this heading">#</a></h2>
<p>By the end of this activity, you will be able to:</p>
<ul class="simple">
<li><p>Apply basic math in Python to summarize an experimental result.</p></li>
<li><p>Create f-strings to display dynamic results.</p></li>
<li><p>Compare and contrast the Nanopore and PacBio systems in terms of sample preparation for nCov2 sequencing.</p></li>
</ul>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="protocol-evaluation">
<h1>Protocol Evaluation<a class="headerlink" href="#protocol-evaluation" title="Link to this heading">#</a></h1>
<p>In the COVID sequencing lab we often have to evaluate multiple techniques for preparing viral samples for sequencing.
There are <em>many</em> aspects to this comparison, but one important one is the yield of the reaction that prepares DNA for sequencing.
This reaction is called an RT-PCR reaction because it <em>reverse-transcribes</em> the nCov2 RNA into DNA through a <em>polymerase chain reaction</em>.</p>
<p>Currently, our lab uses a system with over 200 pairs of overlapping PCR primers that generate DNA fragments that are, on average, <strong>285 basepairs</strong> long.
We’ll call this the <em>Paragon</em> system.
Short RT-PCR fragments are ideal when dealing with degraded RNA because it allows for the amplification of even shredded RNA.
However, because no fragment is longer than ~300bp it is difficult to tell whether a mutation in one gene is ‘linked’ with a mutation in another gene.</p>
<p>So, we are exploring another technique, which we’ll call the <em>PacBio</em> system.
This system uses 35 overlapping fragments that have an average length of <strong>2200 bp</strong>.
These longer fragments help with understanding the linkage between more distant locations but are more difficult to run on degraded samples.</p>
<p>In this lab, you will work through a basic exercise in evaluating the yield of the <em>Paragon</em> and <em>PacBio</em> systems.</p>
<section id="q1-extract-the-relevant-information-from-the-text-above">
<h2>Q1: Extract the relevant information from the text above<a class="headerlink" href="#q1-extract-the-relevant-information-from-the-text-above" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># It is often useful to define all of your variables at the beginning.</span>
<span class="c1"># It is helpful to include units using &#39;#&#39; to keep track of your calculations</span>

<span class="n">dna_weight</span> <span class="o">=</span> <span class="mi">650</span>  <span class="c1"># g/mole/bp</span>

<span class="c1"># bp</span>
<span class="n">paragon_amplicon_length</span> <span class="o">=</span> <span class="o">...</span>
<span class="n">pacbio_amplicon_length</span> <span class="o">=</span> <span class="o">...</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">grader</span><span class="o">.</span><span class="n">check</span><span class="p">(</span><span class="s2">&quot;q1_amp_length&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="q2-calculate-the-molecular-weight-of-each-template">
<h2>Q2: Calculate the molecular weight of each template<a class="headerlink" href="#q2-calculate-the-molecular-weight-of-each-template" title="Link to this heading">#</a></h2>
<p>In this lab, we are comparing the yield of two different RT-PCR methods for preparing viral samples for nanopore sequencing.
In order to accurately compare the yield of each method, it is important to know the molecular weight of the templates being used.
Therefore, as a part of this lab, we need to calculate the molecular weight of each template.
This will allow us to accurately compare the yield of each method and understand how the different characteristics of each method may impact the overall results.</p>
<p><em>Feeling stuck?</em></p>
<p>Try doing it on <strong>paper</strong> first.</p>
<p><em>Points:</em> 4</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">paragon_template_weight</span> <span class="o">=</span> <span class="o">...</span>
<span class="n">pacbio_template_weight</span> <span class="o">=</span> <span class="o">...</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Complete the cell above before running this one</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The Paragon templates weigh </span><span class="si">{</span><span class="n">paragon_template_weight</span><span class="si">}</span><span class="s1"> g/mole&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The PacBio templates weigh </span><span class="si">{</span><span class="n">pacbio_template_weight</span><span class="si">}</span><span class="s1"> g/mole&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">grader</span><span class="o">.</span><span class="n">check</span><span class="p">(</span><span class="s2">&quot;q2_mol_weight&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>In order to investigate the impact of degradation on the yield of each protocol, you will examine two samples.
One sample has been freshly isolated and the other has been left at room temperature for 72 hours before preparation.
We then ran each sample according to the manufacturer’s guidelines.</p>
<p>We obtained the following results after quantification of <strong>15 ul</strong> of final volume.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>sample</p></th>
<th class="head"><p>Fresh</p></th>
<th class="head"><p>Degraded</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Paragon</p></td>
<td><p>21.4 ng/ul</p></td>
<td><p>19.3 ng/ul</p></td>
</tr>
<tr class="row-odd"><td><p>PacBio</p></td>
<td><p>38.1 ng/ul</p></td>
<td><p>7.4 ng/ul</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="q3-what-is-the-molarity-of-each-paragon-sample">
<h2>Q3: What is the <em>molarity</em> of each <em>Paragon</em> sample?<a class="headerlink" href="#q3-what-is-the-molarity-of-each-paragon-sample" title="Link to this heading">#</a></h2>
<p><em>Feeling stuck?</em></p>
<p>Try doing it on <strong>paper</strong> first.</p>
<p>A typical molarity for this protocol is between 50 and 500 fmols/ul.</p>
<p><em>Points:</em> 4</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add variables to hold the concentration of each sample</span>

<span class="c1"># Answer in fmoles/ul</span>
<span class="n">paragon_fresh_molarity</span> <span class="o">=</span> <span class="o">...</span>
<span class="n">paragon_degraded_molarity</span> <span class="o">=</span> <span class="o">...</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># If the cell above is correct, this will print the results.</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The fresh Paragon-prepped sample has a concentration of </span><span class="si">{</span><span class="n">paragon_fresh_molarity</span><span class="si">:</span><span class="s1">0.1f</span><span class="si">}</span><span class="s1"> fmoles/ul&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The degraded Paragon-prepped sample has a concentration of </span><span class="si">{</span><span class="n">paragon_degraded_molarity</span><span class="si">:</span><span class="s1">0.1f</span><span class="si">}</span><span class="s1"> fmoles/ul&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">grader</span><span class="o">.</span><span class="n">check</span><span class="p">(</span><span class="s2">&quot;q3_molarity&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="q4-what-is-the-yield-of-each-pacbio-sample">
<h2>Q4: What is the yield of each <em>PacBio</em> sample?<a class="headerlink" href="#q4-what-is-the-yield-of-each-pacbio-sample" title="Link to this heading">#</a></h2>
<p>Now, let’s calculate the yield of the fresh and degraded PacBio samples.
The yield is the total amount of DNA produced in the reaction (as measured in femptomoles).
This information is important because it allows us to compare the efficiency of the PacBio system in both fresh and degraded samples and determine whether it is a suitable method for our purposes.</p>
<p><em>Points:</em> 5</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add variables to hold the concentration of each sample in ng/ul</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate the molarity of each sample</span>

<span class="n">pacbio_fresh_molarity</span> <span class="o">=</span> <span class="o">...</span>
<span class="n">pacbio_degraded_molarity</span> <span class="o">=</span> <span class="o">...</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The fresh PacBio-prepped sample had a molarity of </span><span class="si">{</span><span class="n">pacbio_fresh_molarity</span><span class="si">:</span><span class="s1">0.1f</span><span class="si">}</span><span class="s1"> fmoles/ul&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The degraded PacBio-prepped sample had a molarity of </span><span class="si">{</span><span class="n">pacbio_degraded_molarity</span><span class="si">:</span><span class="s1">0.1f</span><span class="si">}</span><span class="s1"> fmoles/ul&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate the total vield</span>

<span class="n">pacbio_fresh_yield</span> <span class="o">=</span> <span class="o">...</span>
<span class="n">pacbio_degraded_yield</span> <span class="o">=</span> <span class="o">...</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The fresh PacBio-prepped sample had a yield of </span><span class="si">{</span><span class="n">pacbio_fresh_yield</span><span class="si">:</span><span class="s1">0.1f</span><span class="si">}</span><span class="s1"> fmoles&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The degraded PacBio-prepped sample has a yield of </span><span class="si">{</span><span class="n">pacbio_degraded_yield</span><span class="si">:</span><span class="s1">0.1f</span><span class="si">}</span><span class="s1"> fmoles&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">grader</span><span class="o">.</span><span class="n">check</span><span class="p">(</span><span class="s2">&quot;q4_dna_yield&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="q5-which-samples-are-usable">
<h2>Q5: Which samples are <em>usable</em>?<a class="headerlink" href="#q5-which-samples-are-usable" title="Link to this heading">#</a></h2>
<p>In order to determine which samples are suitable for use in the sequencing protocol, you must first determine if they have sufficient concentration of DNA.
As stated in the protocol, you need to provide 200 fmoles of DNA in 10 ul of dH20.
Based on the yield data that you calculated for the fresh and degraded PacBio and Paragon samples, you will need to determine which samples have a concentration of DNA that meets or exceeds this requirement.
In other words, you will need to determine which samples are usable for the sequencing protocol based on their DNA concentration.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Do the necessary calculations in this cell or the create more above.</span>
<span class="c1"># Remember, you can use variables that you have calculated in cells before.</span>
<span class="c1"># Finally, write a set of formatted print statements that describe the volume of DNA required for each reaction.</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># put a &#39;yes&#39; or &#39;no&#39; for each sample</span>

<span class="n">paragon_fresh_usable</span> <span class="o">=</span> <span class="o">...</span>
<span class="n">paragon_degraded_usable</span> <span class="o">=</span> <span class="o">...</span>

<span class="n">pacbio_fresh_usable</span> <span class="o">=</span> <span class="o">...</span>
<span class="n">pacbio_degraded_usable</span> <span class="o">=</span> <span class="o">...</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">grader</span><span class="o">.</span><span class="n">check</span><span class="p">(</span><span class="s2">&quot;q5_usable_samples&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="conclusion">
<h1>Conclusion<a class="headerlink" href="#conclusion" title="Link to this heading">#</a></h1>
<p>Congratulations on completing this lab! You have successfully compared the yield of two different RT-PCR methods for preparing viral samples for genomic sequencing.
You have determined that the Nanopore system is ideal for degraded samples due to its shorter PCR fragments, while the PacBio system is better for fresher samples and is able to produce longer sequencing fragments.
This is just the beginning of your journey with Python and data analysis.
Keep up the great work as we continue to expand on these techniques and explore more complicated results in future assignments.</p>
<hr class="docutils" />
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./_bblearn/Module02"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="Module02_walkthrough_SOLUTION.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Walkthrough</p>
      </div>
    </a>
    <a class="right-next"
       href="../../content/Module03/Module03_book.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Module 3: DataFrames</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Lab</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-objectives">Learning Objectives</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#protocol-evaluation">Protocol Evaluation</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#q1-extract-the-relevant-information-from-the-text-above">Q1: Extract the relevant information from the text above</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#q2-calculate-the-molecular-weight-of-each-template">Q2: Calculate the molecular weight of each template</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#q3-what-is-the-molarity-of-each-paragon-sample">Q3: What is the <em>molarity</em> of each <em>Paragon</em> sample?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#q4-what-is-the-yield-of-each-pacbio-sample">Q4: What is the yield of each <em>PacBio</em> sample?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#q5-which-samples-are-usable">Q5: Which samples are <em>usable</em>?</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#conclusion">Conclusion</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Will Dampier
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>